### PR TITLE
[SPARK-8970][SQL] remove unnecessary abstraction for ExtractValue

### PR DESCRIPTION
**Pull request by cloud-fan**

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @rxin

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala [hunk 1]
```diff
@@ -78,10 +78,12 @@ object ExtractValue {
     }
   }
 
-  def unapply(g: ExtractValue): Option[(Expression, Expression)] = g match {
+  def unapply(g: Expression): Option[(Expression, Expression)] = g match {
     case o: GetArrayItem => Some((o.child, o.ordinal))
     case o: GetMapValue => Some((o.child, o.key))
-    case s: ExtractValueWithStruct => Some((s.child, null))
+    case o: GetStructField => Some((o.child, null))
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala]: I find it odd to return null in the 2nd field. Why is the semantics of that?

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala]: This `unapply` is only used in optimizer to save code. I have removed it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -275,8 +275,12 @@ object NullPropagation extends Rule[LogicalPlan] {
...
       case e @ IsNull(c) if !c.nullable => Literal.create(false, BooleanType)
       case e @ IsNotNull(c) if !c.nullable => Literal.create(true, BooleanType)
-      case e @ ExtractValue(Literal(null, _), _) => Literal.create(null, e.dataType)
-      case e @ ExtractValue(_, Literal(null, _)) => Literal.create(null, e.dataType)
+      case e @ GetArrayItem(Literal(null, _), _) => Literal.create(null, e.dataType)
+      case e @ GetArrayItem(_, Literal(null, _)) => Literal.create(null, e.dataType)
+      case e @ GetMapValue(Literal(null, _), _) => Literal.create(null, e.dataType)
+      case e @ GetMapValue(_, Literal(null, _)) => Literal.create(null, e.dataType)
+      case e @ GetStructField(Literal(null, _), _, _) => Literal.create(null, e.dataType)
+      case e @ GetArrayStructFields(Literal(null, _), _, _) => Literal.create(null, e.dataType)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Maybe we can simplify null propagation for binary and unary expression. Will do that in another PR.

**Comment by rxin** [comment]
Jenkins, retest this please.

**Comment by rxin** [comment]
Jenkins, retest this please.

**Comment by cloud-fan** [comment]
I saw the JIRA is marked as resolved, any problems here?

**Comment by rxin** [comment]
It's been merged -- but github didn't close this. Can you close the pr?
